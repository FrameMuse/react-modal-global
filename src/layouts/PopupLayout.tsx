/*

MIT License

Copyright (c) 2023 Valery Zinchenko

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

*/

import { ReactNode } from "react"

import { useModalContext } from "../hooks"

interface PopupLayoutProps {
  width?: string;
  children: ReactNode;
}

function PopupLayout(props: PopupLayoutProps) {
  const modal = useModalContext()

  return (
    <div className="popup-layout" style={{ "--popup-width": props.width }}>
      <div className="popup-layout__container">
        {modal.params.closable && (
          <button
            className="popup-layout__close"
            type="button"
            onClick={modal.close}
          >
            &#9587;
          </button>
        )}
        <div className="popup-layout__inner">{props.children}</div>
      </div>
    </div>
  )
}

export default PopupLayout
