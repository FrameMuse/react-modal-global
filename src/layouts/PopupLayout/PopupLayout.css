/*

MIT License

Copyright (c) 2023 Valery Zinchenko

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

*/

.popup-layout {
  position: relative;

  padding: 2em;
  background: white;
  border-radius: 1em;

  min-width: 20em;
  max-width: 75vw;

  max-height: 90vh;

  margin: auto;
  overflow: auto;
  overflow: overlay;
  overscroll-behavior: contain;

  color: #333;
  box-shadow: 0px 0px 20px 10px rgba(black, 0.1);
  cursor: initial;

  transition: 250ms ease top;
  top: -2.5em;
}

.modal--active .popup-layout {
  top: 0em;
}

.popup-layout__close {
  margin-left: auto;

  position: absolute;
  top: 1.25em;
  right: 1.5em;
}

.popup-layout__container {
  display: grid;
  gap: 2.5em;
}