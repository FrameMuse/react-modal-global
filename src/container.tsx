import React from "react"
import { classWithModifiers } from "./utils"
import { PopupContext } from "./context"
import { PopupWindow } from "./interfaces"
import { PopupPrivate } from "./controller"

export interface PopupContainerProps {
  className?: string
}
export interface PopupContainerState {
  isActive: boolean
  queue: PopupWindow[]
}

export class PopupContainer extends React.Component<PopupContainerProps, PopupContainerState> {
  state: PopupContainerState = {
    isActive: false,
    queue: []
  }

  constructor(props: any) {
    super(props)
    // Set Popup dispatcher
    PopupPrivate.dispatch = this.setState.bind(this)
  }

  render() {
    const { isActive, queue } = this.state
    const lastPopup = queue[queue.length - 1]
    const { component: PopupWindowComponent, params = {}, close } = lastPopup || {}

    const className = this.props.className || "popup"
    return (
      <div className={classWithModifiers(className, isActive && "active")}>
        <div className={className + "__container"} onClick={close}>
          <div className={className + "__inner"} onClick={event => event.stopPropagation()}>
            <PopupContext.Provider value={lastPopup}>
              {PopupWindowComponent && <PopupWindowComponent {...params} />}
            </PopupContext.Provider>
          </div>
        </div>
      </div>
    )
  }
}
