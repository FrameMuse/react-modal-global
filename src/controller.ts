import { Dispatch, SetStateAction } from "react";
import { PopupContainerState } from "./container";
import { PopupComponent, PopupParams, PopupWindow } from "./interfaces"

type AnyIfEmpty<T extends object> = keyof T extends never ? any : T;

export const PopupPrivate: {
  dispatch: Dispatch<SetStateAction<PopupContainerState>>
} = {
  dispatch: () => { }
}

export class Popup {
  public static open
    <P extends object = {}, AC extends Partial<PopupParams> & P = Partial<PopupParams> & P>
    (component: PopupComponent<P>, ...[params]: (AnyIfEmpty<P> extends object ? [AC] : [AC?])): Promise<void> {
    return new Promise<void>(function (resolve) {
      const popupWindow = { component, params, close }
      Popup.addToQueue(popupWindow)
      function close() {
        resolve()
        Popup.removeFromQueue(popupWindow)
      }
    })
  }
  private static addToQueue(popupWindow: PopupWindow<any>) {
    PopupPrivate.dispatch(state => ({
      isActive: true,
      queue: [...state.queue, popupWindow]
    }))
  }
  private static removeFromQueue(popupWindow: PopupWindow<any>) {
    PopupPrivate.dispatch(state => {
      const queue = state.queue.filter(pw => pw !== popupWindow)
      return {
        isActive: queue.length > 0,
        queue
      }
    })
  }
  public static closeAll() {
    PopupPrivate.dispatch(state => {
      state.queue.forEach(popup => popup.close())
      return {
        isActive: false,
        queue: []
      }
    })
  }
}
