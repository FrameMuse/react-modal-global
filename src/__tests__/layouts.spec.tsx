import { render } from "@testing-library/react"
import { act } from "react-dom/test-utils"

import { ModalContainer } from "../container"
import { ModalController } from "../controller"
import DrawerLayout from "../layouts/DrawerLayout/DrawerLayout"
import PopupLayout from "../layouts/PopupLayout/PopupLayout"

let Modal: ModalController = new ModalController()
let container = render(<ModalContainer controller={Modal} />).container
let modalContainerElement = container.querySelector(".modal")

beforeEach(() => {
  Modal = new ModalController()
  container = render(<ModalContainer controller={Modal} />).container
  modalContainerElement = container.querySelector(".modal")
})

describe("Layouts", () => {
  describe("PopupLayout", () => {
    it("should render", () => {
      act(() => {
        Modal.open(() => <PopupLayout>123</PopupLayout>)
      })

      expect(modalContainerElement).not.toBeNull()
      expect(modalContainerElement).toMatchSnapshot()
    })
  })

  describe("DrawerLayout", () => {
    it("should render", () => {
      act(() => {
        Modal.open(() => <DrawerLayout>123</DrawerLayout>)
      })

      expect(modalContainerElement).not.toBeNull()
      expect(modalContainerElement).toMatchSnapshot()
    })
  })
})