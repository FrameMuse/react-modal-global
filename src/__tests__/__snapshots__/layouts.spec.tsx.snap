// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layouts DrawerLayout should render 1`] = `
<div
  aria-hidden="false"
  aria-modal="true"
  class="modal modal--active"
>
  <div
    class="modal__container"
  >
    <div
      class="drawer-layout"
    >
      <div
        class="drawer-layout__container"
      >
        <button
          class="drawer-layout__close"
          type="button"
        >
          ╳
        </button>
        <div
          class="drawer-layout__inner"
        >
          123
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Layouts PopupLayout should render 1`] = `
<div
  aria-hidden="false"
  aria-modal="true"
  class="modal modal--active"
>
  <div
    class="modal__container"
  >
    <div
      class="popup-layout"
    >
      <div
        class="popup-layout__container"
      >
        <button
          class="popup-layout__close"
          type="button"
        >
          ╳
        </button>
        <div
          class="popup-layout__inner"
        >
          123
        </div>
      </div>
    </div>
  </div>
</div>
`;
